### ISSUE TITLE
defunct processes after every minion task in docker container

### ISSUE THREAD

**Issue by maci0** [saltstack/salt#9800]
Running salt-minion in a docker container.

```
root@test /home/maci/minion # docker run -d minion
e8091476d9f2eb836e3f896c4f2bdf93ff46c537e5c5ec55ef62be110ad06fcd
root@test /home/maci/minion # ps aux|grep salt-minion 
root     30879  0.2  0.1  18904  1152 ?        Ss   12:14   0:00 lxc-start -n e8091476d9f2eb836e3f896c4f2bdf93ff46c537e5c5ec55ef62be110ad06fcd -f /var/lib/docker/containers/e8091476d9f2eb836e3f896c4f2bdf93ff46c537e5c5ec55ef62be110ad06fcd/config.lxc -- /.dockerinit -g 172.17.42.1 -i 172.17.0.10/16 -- /usr/bin/salt-minion
root     30885 74.7  2.1 335220 22016 ?        S    12:14   0:02 /usr/bin/python /usr/bin/salt-minion
root     30954  0.0  0.0 103244   844 pts/0    S+   12:14   0:00 grep --color=auto salt-minion
root@test /home/maci/minion # salt-key -A
The following keys are going to be accepted:
Unaccepted Keys:
e8091476d9f2
Proceed? [n/Y] y
Key for minion e8091476d9f2 accepted.
root@test /home/maci/minion # ps aux|grep salt-minion
root     30879  0.0  0.1  18904  1152 ?        Ss   12:14   0:00 lxc-start -n e8091476d9f2eb836e3f896c4f2bdf93ff46c537e5c5ec55ef62be110ad06fcd -f /var/lib/docker/containers/e8091476d9f2eb836e3f896c4f2bdf93ff46c537e5c5ec55ef62be110ad06fcd/config.lxc -- /.dockerinit -g 172.17.42.1 -i 172.17.0.10/16 -- /usr/bin/salt-minion
root     30885 24.9  2.1 335220 22016 ?        S    12:14   0:02 /usr/bin/python /usr/bin/salt-minion
root     31228  0.0  0.0 103244   848 pts/0    R+   12:14   0:00 grep --color=auto salt-minion
root@test /home/maci/minion # salt 
1 root@test /home/maci/minion # salt '*' test.ping                                                             :(
e8091476d9f2:
    True
root@test /home/maci/minion # salt '*' test.ping
e8091476d9f2:
    True
root@test /home/maci/minion # salt '*' test.ping
e8091476d9f2:
    True
root@test /home/maci/minion # ps aux|grep salt-minion
root     30879  0.0  0.1  18904  1152 ?        Ss   12:14   0:00 lxc-start -n e8091476d9f2eb836e3f896c4f2bdf93ff46c537e5c5ec55ef62be110ad06fcd -f /var/lib/docker/containers/e8091476d9f2eb836e3f896c4f2bdf93ff46c537e5c5ec55ef62be110ad06fcd/config.lxc -- /.dockerinit -g 172.17.42.1 -i 172.17.0.10/16 -- /usr/bin/salt-minion
root     30885 17.5  3.7 546608 38316 ?        Sl   12:14   0:03 /usr/bin/python /usr/bin/salt-minion
root     31254  0.0  0.0      0     0 ?        Zs   12:14   0:00 [salt-minion] <defunct>
root     31255  2.0  0.0      0     0 ?        Z    12:14   0:00 [salt-minion] <defunct>
root     31269  0.0  0.0      0     0 ?        Zs   12:14   0:00 [salt-minion] <defunct>
root     31270  0.0  0.0      0     0 ?        Z    12:14   0:00 [salt-minion] <defunct>
root     31284  0.0  0.0      0     0 ?        Zs   12:14   0:00 [salt-minion] <defunct>
root     31285  0.0  0.0      0     0 ?        Z    12:14   0:00 [salt-minion] <defunct>
root     31299  0.0  0.0      0     0 ?        Zs   12:14   0:00 [salt-minion] <defunct>
root     31300  0.0  0.0      0     0 ?        Z    12:14   0:00 [salt-minion] <defunct>
root     31306  0.0  0.0 103244   840 pts/0    R+   12:14   0:00 grep --color=auto salt-minion
```

**Comment by mavenAtHouzz** [comment]
_Hope this can help those who still waiting on this issue update._

This is not the salt issue but a typical pid 1 reaping children processes problem when running docker. Can't find whether docker resolved it (to my env "Docker version 1.10.2, build c3959b1" still leave it to the client to solve it). Here's the detailed [explanation](https://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/).

I prefer the quick bash fix instead of adding the more complicated minimized python init process (that is a more generic solution but for salt-minion bash is enough in my opinion). Just start the minion container with the following wrapped script.

```shell
#!/bin/bash

function cleanup() {
  local pids=`jobs -p`
  if [ "$pids" != "" ]; then
    kill $pids >/dev/null 2>/dev/null
    sleep 5
  fi
}

trap cleanup EXIT
/usr/bin/salt-minion
```

**Comment by hinnerk** [comment]
We've found [dumb-init](http://engineeringblog.yelp.com/2016/01/dumb-init-an-init-for-docker.html) to be a more comprehensive solution.
